common/src/main/java/io/netty/util/concurrent/ImmediateEventExecutor.java

```diff
@@ -15,63 +15,69 @@
  */
 package io.netty.util.concurrent;
 
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
 import java.util.concurrent.TimeUnit;
 
 /**
- * {@link AbstractEventExecutor} which execute tasks in the callers thread.
+ * {@link AbstractEventExecutor} which execute tasks in the caller's thread.
+ * <p>
+ * This class does not provide any protection against re-entry or {@link StackOverflowError}.
+ * Use {@link ReentrantImmediateEventExecutor} if these protections are necessary.
  */
-public final class ImmediateEventExecutor extends AbstractEventExecutor {
+public class ImmediateEventExecutor extends AbstractEventExecutor {
 
     public static final ImmediateEventExecutor INSTANCE = new ImmediateEventExecutor();
 
     private final Future<?> terminationFuture = new FailedFuture<Object>(
             GlobalEventExecutor.INSTANCE, new UnsupportedOperationException());
 
-    private ImmediateEventExecutor() {
-        // Singleton
-    }
+    ImmediateEventExecutor() { }
 
     @Override
-    public boolean inEventLoop() {
+    public final boolean inEventLoop() {
         return true;
     }
 
     @Override
-    public boolean inEventLoop(Thread thread) {
+    public final boolean inEventLoop(Thread thread) {
         return true;
     }
 
     @Override
-    public Future<?> shutdownGracefully(long quietPeriod, long timeout, TimeUnit unit) {
+    public final Future<?> shutdownGracefully(long quietPeriod, long timeout, TimeUnit unit) {
         return terminationFuture();
     }
 
     @Override
-    public Future<?> terminationFuture() {
+    public final Future<?> terminationFuture() {
         return terminationFuture;
     }
 
     @Override
     @Deprecated
-    public void shutdown() { }
+    public final void shutdown() { }
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1186)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the missing @deprecated Javadoc tag. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)